### PR TITLE
Add example for prefetcher Swift

### PR DESCRIPTION
**Pull request by LeOric32**

### CODE REVIEW

**Review by dashchak — CHANGES_REQUESTED** [review]
Looks like you committed 2 duplicated projects nested in each other. Please fix

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    }
+    
+    // MARK: - Buttons action
+    @IBAction func boundingBoxAction(_ sender: UIButton) {
+        let fetch = segmentedControl.selectedSegmentIndex == 0
+        guard let boundingBox = self.mapView.boundingBox else { return }
+        self.view.isUserInteractionEnabled = false
+        activityView.startAnimating()
+        if fetch {
+            prefetcher.fetchMapDataForBoundingBox(boundingBox, error: nil)
```
Comment (dashchak) [inline map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:63]: Please handle error here and in other places of prefetching.
 i.e. when Area is too big, data will not even start prefetching

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    
+    // method calls, when your progress fetching(downloading) is over.
+    func prefetcher(_ prefetcher: NMAMapDataPrefetcher, didUpdateProgress progress: Float,
+                    forRequestId requestId: Int) {
+        showMessage(title: "Success", message: "This region is fetched.")
+    }
+    
+    // MARK: - Route functions
+    func showRoute() {
+
```
Comment (dashchak) [inline map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:102]: Remove line

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        // START: 4350 Still Creek Dr
+        let hereBurnaby = NMAGeoCoordinates(latitude: Defaults.hereBurnabyLatitude,
+                                            longitude: Defaults.hereBurnabyLongitude)
+        
+        // END: Langley BC
+        let langley = NMAGeoCoordinates(latitude: Defaults.langleyLatitude,
+                                        longitude: Defaults.langleyLongitude)
+        
+        stops.append(hereBurnaby)
+        stops.append(langley)
```
Comment (dashchak) [inline map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:142]: let stops = [hereBurnaby, langley]

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        let routingMode = NMARoutingMode(routingType: .balanced,
+                                         transportMode: .car,
+                                         routingOptions: .avoidTunnel)
+        
+        // Trigger the route calculation
+        router.calculateRoute(withStops: stops ,
+                              routingMode: routingMode)
+        { [weak self] routeResult, error in
+            guard error == .none else {
+                return
```
Comment (dashchak) [inline map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:155]: Please show error to user. Sometime route can not be created

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        if let previousMapRoute = mapRoute {
+            mapView.remove(mapObject:previousMapRoute)
+        }
+        
+        guard let unwrappedRoute = route else {
+            return
+        }
+        
+        mapRoute = NMAMapRoute(unwrappedRoute)
+        mapRoute?.traveledColor = .clear
```
Comment (dashchak) [inline map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:179]: please remove this line as navigation is not supported in this app

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        
+        // set zoom level
+        mapView.zoomLevel = Defaults.zoomLevel
+    }
+    
+    // MARK: - Buttons action
+    @IBAction func boundingBoxAction(_ sender: UIButton) {
+        let fetch = segmentedControl.selectedSegmentIndex == 0
+        guard let boundingBox = self.mapView.boundingBox else { return }
+        self.view.isUserInteractionEnabled = false
```
Comment (dashchak) [inline map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:60]: Can be extracted to function and reused

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift (hunk 1 above)
Comment (LeOric32) [reply map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift]: Done

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift (hunk 2 above)
Comment (LeOric32) [reply map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift]: Done

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift (hunk 3 above)
Comment (LeOric32) [reply map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift]: Done

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift (hunk 5 above)
Comment (LeOric32) [reply map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift]: Done

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift (hunk 6 above)
Comment (LeOric32) [reply map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift]: Done

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift (hunk 4 above)
Comment (LeOric32) [reply map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:175]: Done

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        
+        static let hereBurnabyLatitude  = 49.264178
+        static let hereBurnabyLongitude = -123.117542
+        
+        static let langleyLatitude  = 49.229185
+        static let langleyLongitude = -123.094052
+        
+        static let defaultRadius: UInt = 600
+        
+        static let frame = CGRect(x: 110, y: 200, width: 220, height: 120)
```
Comment (dashchak) [inline map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:24]: Remove it

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        activityView.stopAnimating()
+        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
+        alert.addAction(UIAlertAction(title: "Ok", style: .default, handler: nil))
+        self.present(alert, animated: true, completion: nil)
+        self.view.isUserInteractionEnabled = true
+    }
+}
+
+extension ViewController: NMAMapViewDelegate {
+    func mapViewDidDraw(_ mapView: NMAMapView) {
```
Comment (dashchak) [inline map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:219]: Is not this callback called many times? You can create route in viewDidAppear

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+            self.showMessage(title: "Error", message: "Can't get bounding box.")
+            return
+        }
+        activityProcess()
+        
+        var error = NMAPrefetchRequestError.unknown
+        prefetcher.estimateMapDataSizeForBoundingBox(boundingBox, error: &error)
+        
+        if error != .none {
+            self.showMessage(title: "Error", message: "Can't estimate area.")
```
Comment (dashchak) [inline map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:68]: Also include error number

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+            self.showMessage(title: "Error", message: "Can't get bounding box.")
+            return
+        }
+        activityProcess()
+        
+        var error = NMAPrefetchRequestError.unknown
+        prefetcher.fetchMapDataForBoundingBox(boundingBox, error: &error)
+        
+        if error != .none {
+            self.showMessage(title: "Error", message: "Can't fetch area.")
```
Comment (dashchak) [inline map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:84]: Show error number

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        }
+        showRoute()
+    }
+    
+    // MARK: - NMAMapDataPrefetcherListener
+    func prefetcher(_ prefetcher: NMAMapDataPrefetcher, didEstimateSize dataSizeKB: Int,
+                    forRequestId requestId: Int, withSuccess success: Bool) {
+        let message = dataSizeKB > 0 ? "Estimated size for this region is \(dataSizeKB) KB"
+            : "This region is already fetched."
+        showMessage(title: "Success", message: message)
```
Comment (dashchak) [inline map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:128]: use title depending on "withSuccess success"

File: map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+        // set map view with geo center
+        mapView.set(geoCenter: geoCoordCenter, animation: .none)
+        
+        // set zoom level
+        mapView.zoomLevel = Defaults.zoomLevel
+    }
+    
+    // MARK: - Buttons action
+    @IBAction func createRoute(_ sender: UIButton) {
+        self.createRoute()
```
Comment (dashchak) [inline map-data-prefetcher-ios-swift/MapDataPrefetcherApp/ViewController.swift:54]: Not needed to use extra method for it

**Pull request closed without merging**